jsonschema/src/compilation/mod.rs

```diff
@@ -302,4 +320,253 @@ mod tests {
         );
         assert_eq!(errors[1].to_string(), r#""a" is shorter than 3 characters"#);
     }
+
+    #[test]
+    fn custom_keyword_definition() {
+        // Define a custom validator that verifies the object's keys consist of
+        // only ASCII representable characters.
+        #[allow(clippy::needless_pass_by_value)]
+        fn custom_object_validator(
+            instance: &Value,
+            instance_path: JSONPointer,
+            schema: Arc<Value>,
+            schema_path: JSONPointer,
+            _: Arc<Value>,
+        ) -> ErrorIterator<'_> {
+            if schema.as_str().map_or(true, |str| str != "ascii-keys") {
+                let error = ValidationError {
+                    instance: Cow::Borrowed(instance),
+                    kind: crate::error::ValidationErrorKind::Schema,
+                    instance_path,
+                    schema_path,
+                };
+                return Box::new(Some(error).into_iter()); // Invalid schema
+            }
+            let mut errors = vec![];
+            for (key, _value) in instance.as_object().unwrap() {
+                if !key.is_ascii() {
+                    let error = ValidationError {
+                        instance: Cow::Borrowed(instance),
+                        kind: crate::error::ValidationErrorKind::Format { format: "ASCII" },
+                        instance_path: instance_path.clone(),
+                        schema_path: schema_path.clone(),
+                    };
+                    errors.push(error);
+                }
+            }
+            Box::new(errors.into_iter())
+        }
+        fn is_custom_object_valid(instance: &Value, schema: &Value, _: &Value) -> bool {
+            if schema.as_str().map_or(true, |str| str != "ascii-keys") {
+                return false; // Invalid schema
+            }
+            for (key, _value) in instance.as_object().unwrap() {
+                if !key.is_ascii() {
+                    return false;
+                }
+            }
+            true
+        }
+        let definition = CustomKeywordDefinition::Validator {
```

Comment:
That sounds like a good idea, but a few clarifying questions: do you mean creating a trait like `CustomValidator` with the `validate` and `is_valid` functions as methods on it, and then the user would define a struct that implements `CustomValidator` and provide it to CompilationOptions (in like a Box I think)? If so perhaps it should be combined somehow with the `Validator` trait?

Also while instinctually this feels like the right move, I'm struggling a bit to come up with a concrete use case example, like for a test case. There's this https://github.com/Stranger6667/jsonschema-rs/pull/394#issuecomment-2047665476, but I wonder if that doesn't need state _internal_ to the validator, the `validate` / `is_valid` functions could potentially just reference the lookup table. Something like that's probably worth a test case anyway also.

Reply:
> do you mean creating a trait like CustomValidator with the validate and is_valid functions as methods on it, and then the user would define a struct that implements CustomValidator and provide it to CompilationOptions (in like a Box I think)? If so perhaps it should be combined somehow with the Validator trait?

Yep, or, maybe exposing the `Validator` trait?

> Also while instinctually this feels like the right move, I'm struggling a bit to come up with a concrete use case example, like for a test case

An example that I have in mind is a counter. Even though it might sound trivial, it unblocks interesting use cases like collecting usage statistics from different keywords (i.e. if the user "wraps" the existing keyword implementation with such a counter), which in turn is useful for building coverage maps for test cases send to API endpoints (documented with Open API), which in turn is useful for coverage-guided fuzzing of such endpoints.